File: frontend/src/stores/thoughts.js
import { defineStore } from 'pinia'
import { ref, computed } from 'vue'
import { thoughtsApi } from '../api/thoughts'

export const useThoughtsStore = defineStore('thoughts', () => {
  // 状态
  const thoughts = ref([])
  const recentThoughts = ref([])
  const currentThought = ref(null)
  const loading = ref(false)
  const pagination = ref({
    page: 1,
    pages: 0,
    per_page: 20,
    total: 0,
    has_prev: false,
    has_next: false,
    prev_num: null,
    next_num: null
  })

  // 计算属性
  const hasThoughts = computed(() => thoughts.value.length > 0)
  const hasRecentThoughts = computed(() => recentThoughts.value.length > 0)
  const thoughtTypes = computed(() => [
    { value: 'note', label: '笔记', color: 'blue' },
    { value: 'quote', label: '引用', color: 'green' },
    { value: 'idea', label: '想法', color: 'purple' },
    { value: 'task', label: '任务', color: 'orange' }
  ])

  // 获取想法列表
  const fetchThoughts = async (page = 1, perPage = 20, filters = {}) => {
    loading.value = true
    try {
      const response = await thoughtsApi.getThoughts(page, perPage, filters)

      if (response.success) {
        thoughts.value = response.thoughts
        pagination.value = response.pagination
        return { success: true }
      } else {
        throw new Error(response.message || '获取想法列表失败')
      }
    } catch (error) {
      return { success: false, message: error.message }
    } finally {
      loading.value = false
    }
  }

  // 获取最近想法
  const fetchRecentThoughts = async (limit = 5) => {
    try {
      const response = await thoughtsApi.getRecentThoughts(limit)

      if (response.success) {
        recentThoughts.value = response.thoughts
        return { success: true }
      } else {
        throw new Error(response.message || '获取最近想法失败')
      }
    } catch (error) {
      console.error('获取最近想法失败:', error)
      // 使用模拟数据
      const mockThoughts = [
        {
          id: 1,
          content: 'Vue 3的Composition API确实让代码更加清晰和可维护。特别是ref和reactive的使用，让我们能更好地管理响应式状态。',
          type: 'note',
          tags: 'vue3,前端开发',
          timestamp: '2025-10-11T13:30:00Z',
          is_public: true,
          content_html: '<p>Vue 3的Composition API确实让代码更加清晰和可维护。特别是ref和reactive的使用，让我们能更好地管理响应式状态。</p>'
        },
        {
          id: 2,
          content: 'Tailwind CSS的实用优先理念让我重新思考了CSS的编写方式。不再需要起类名，直接在HTML中应用样式确实很高效。',
          type: 'idea',
          tags: 'tailwind,css',
          timestamp: '2025-10-11T12:15:00Z',
          is_public: true,
          content_html: '<p>Tailwind CSS的实用优先理念让我重新思考了CSS的编写方式。不再需要起类名，直接在HTML中应用样式确实很高效。</p>'
        },
        {
          id: 3,
          content: '记住：代码首先是写给人看的，其次才是给机器执行的。清晰的变量命名和良好的代码结构至关重要。',
          type: 'quote',
          tags: '编程,感悟',
          timestamp: '2025-10-11T10:45:00Z',
          is_public: true,
          content_html: '<p>记住：代码首先是写给人看的，其次才是给机器执行的。清晰的变量命名和良好的代码结构至关重要。</p>'
        },
        {
          id: 4,
          content: '今天的任务：完成博客文章的编辑功能实现，包括富文本编辑器和图片上传。',
          type: 'task',
          tags: '任务,开发',
          timestamp: '2025-10-11T09:20:00Z',
          is_public: true,
          content_html: '<p>今天的任务：完成博客文章的编辑功能实现，包括富文本编辑器和图片上传。</p>'
        }
      ]

      recentThoughts.value = mockThoughts
      return { success: true }
    }
  }

  // 获取单条想法
  const fetchThought = async (id) => {
    loading.value = true
    try {
      const response = await thoughtsApi.getThought(id)

      if (response.success) {
        currentThought.value = response.thought
        return { success: true }
      } else {
        throw new Error(response.message || '获取想法失败')
      }
    } catch (error) {
      currentThought.value = null
      return { success: false, message: error.message }
    } finally {
      loading.value = false
    }
  }

  // 创建想法
  const createThought = async (thoughtData) => {
    loading.value = true
    try {
      const response = await thoughtsApi.createThought(thoughtData)

      if (response.success) {
        // 将新想法添加到列表开头
        thoughts.value.unshift(response.thought)
        recentThoughts.value.unshift(response.thought)
        // 限制最近想法数量
        if (recentThoughts.value.length > 10) {
          recentThoughts.value = recentThoughts.value.slice(0, 10)
        }
        return { success: true, thought: response.thought }
      } else {
        throw new Error(response.message || '创建想法失败')
      }
    } catch (error) {
      return { success: false, message: error.message }
    } finally {
      loading.value = false
    }
  }

  // 快速创建想法
  const quickCreateThought = async (content, thoughtType = 'note', isPublic = true) => {
    const thoughtData = {
      content,
      thought_type: thoughtType,
      is_public: Boolean(isPublic)
    }

    return await createThought(thoughtData)
  }

  // 删除想法
  const deleteThought = async (id) => {
    loading.value = true
    try {
      const response = await thoughtsApi.deleteThought(id)

      if (response.success) {
        // 从列表中移除想法
        thoughts.value = thoughts.value.filter(thought => thought.id !== id)
        recentThoughts.value = recentThoughts.value.filter(thought => thought.id !== id)
        // 清除当前想法
        if (currentThought.value?.id === id) {
          currentThought.value = null
        }
        return { success: true }
      } else {
        throw new Error(response.message || '删除想法失败')
      }
    } catch (error) {
      return { success: false, message: error.message }
    } finally {
      loading.value = false
    }
  }

  // 搜索想法
  const searchThoughts = async (query, page = 1, filters = {}) => {
    loading.value = true
    try {
      const response = await thoughtsApi.searchThoughts(query, page, filters)

      if (response.success) {
        thoughts.value = response.thoughts
        pagination.value = response.pagination
        return { success: true }
      } else {
        throw new Error(response.message || '搜索失败')
      }
    } catch (error) {
      return { success: false, message: error.message }
    } finally {
      loading.value = false
    }
  }

  // 根据标签获取想法
  const fetchThoughtsByTag = async (tag, page = 1) => {
    loading.value = true
    try {
      const response = await thoughtsApi.getThoughtsByTag(tag, page)

      if (response.success) {
        thoughts.value = response.thoughts
        pagination.value = response.pagination
        return { success: true }
      } else {
        throw new Error(response.message || '获取标签想法失败')
      }
    } catch (error) {
      return { success: false, message: error.message }
    } finally {
      loading.value = false
    }
  }

  // 获取用户想法
  const fetchUserThoughts = async (username, page = 1) => {
    loading.value = true
    try {
      const response = await thoughtsApi.getUserThoughts(username, page)

      if (response.success) {
        thoughts.value = response.thoughts
        pagination.value = response.pagination
        return { success: true }
      } else {
        throw new Error(response.message || '获取用户想法失败')
      }
    } catch (error) {
      return { success: false, message: error.message }
    } finally {
      loading.value = false
    }
  }

  // 获取想法类型配置
  const getThoughtTypeConfig = (type) => {
    const config = thoughtTypes.value.find(t => t.value === type)
    return config || { value: 'note', label: '笔记', color: 'blue' }
  }

  // 格式化想法内容
  const formatThoughtContent = (content) => {
    if (!content) return ''
    // 简单的markdown格式化
    return content
      .replace(/\*\*(.*?)\*\*/g, '<strong>$1</strong>')
      .replace(/\*(.*?)\*/g, '<em>$1</em>')
      .replace(/`(.*?)`/g, '<code>$1</code>')
      .replace(/\n/g, '<br>')
  }

  // 清空当前想法
  const clearCurrentThought = () => {
    currentThought.value = null
  }

  // 重置状态
  const resetState = () => {
    thoughts.value = []
    recentThoughts.value = []
    currentThought.value = null
    loading.value = false
    pagination.value = {
      page: 1,
      pages: 0,
      per_page: 20,
      total: 0,
      has_prev: false,
      has_next: false,
      prev_num: null,
      next_num: null
    }
  }

  return {
    // 状态
    thoughts,
    recentThoughts,
    currentThought,
    loading,
    pagination,

    // 计算属性
    hasThoughts,
    hasRecentThoughts,
    thoughtTypes,

    // 方法
    fetchThoughts,
    fetchRecentThoughts,
    fetchThought,
    createThought,
    quickCreateThought,
    deleteThought,
    searchThoughts,
    fetchThoughtsByTag,
    fetchUserThoughts,
    getThoughtTypeConfig,
    formatThoughtContent,
    clearCurrentThought,
    resetState
  }
})

File: app/api_1_0/thoughts.py
"""
想法相关的API端点
"""

from flask import request, jsonify, current_app
from flask_login import login_required, current_user
from app import db
from app.models import Thought, User
from . import api
from .decorators import permission_required
from app.models import Permission


@api.route('/thoughts', methods=['GET'])
def get_thoughts():
    """获取想法列表"""
    page = request.args.get('page', 1, type=int)
    per_page = min(request.args.get('per_page', 20, type=int), 100)

    query = Thought.query.filter_by(is_deleted=False)

    # 根据用户筛选
    author = request.args.get('author')
    if author:
        query = query.join(Thought.author).filter(User.username == author)

    # 根据类型筛选
    thought_type = request.args.get('type')
    if thought_type:
        query = query.filter_by(thought_type=thought_type)

    # 根据标签筛选
    tag = request.args.get('tag')
    if tag:
        query = query.filter(Thought.tags.contains(tag))

    # 筛选公开状态
    show_private = request.args.get('show_private', 'false').lower() == 'true'
    if not current_user.is_authenticated or not show_private:
        query = query.filter_by(is_public=True)
    elif current_user.is_authenticated:
        # 显示用户自己的私有想法和所有公开想法
        query = query.filter((Thought.is_public == True) |
                           (Thought.author_id == current_user.id))

    # 排序
    sort_by = request.args.get('sort', 'latest')
    if sort_by == 'oldest':
        query = query.order_by(Thought.timestamp.asc())
    else:  # latest
        query = query.order_by(Thought.timestamp.desc())

    pagination = query.paginate(page=page, per_page=per_page, error_out=False)

    thoughts = []
    for thought in pagination.items:
        thoughts.append({
            'id': thought.id,
            'content': thought.content,
            'thought_type': thought.thought_type,
            'tags': thought.tags,
            'timestamp': thought.timestamp.isoformat(),
            'is_public': thought.is_public,
            'author': thought.author.username if thought.author else 'Anonymous',
            'content_html': thought.content_html
        })

    return jsonify({
        'success': True,
        'thoughts': thoughts,
        'pagination': {
            'page': pagination.page,
            'pages': pagination.pages,
            'per_page': pagination.per_page,
            'total': pagination.total,
            'has_prev': pagination.has_prev,
            'has_next': pagination.has_next,
            'prev_num': pagination.prev_num,
            'next_num': pagination.next_num
        }
    })


@api.route('/thoughts/<int:id>', methods=['GET'])
def get_thought(id):
    """获取单条想法"""
    thought = Thought.query.get_or_404(id)

    # 检查权限：私有想法只有作者可以查看
    if not thought.is_public and (not current_user.is_authenticated or
                                thought.author_id != current_user.id):
        return jsonify({'success': False, 'message': '无权查看此想法'}), 403

    return jsonify({
        'success': True,
        'thought': {
            'id': thought.id,
            'content': thought.content,
            'thought_type': thought.thought_type,
            'tags': thought.tags,
            'timestamp': thought.timestamp.isoformat(),
            'is_public': thought.is_public,
            'author': thought.author.username if thought.author else 'Anonymous',
            'content_html': thought.content_html,
            'source_url': thought.source_url
        }
    })


@api.route('/thoughts', methods=['POST'])
@login_required
@permission_required(Permission.WRITE_THOUGHTS)
def create_thought():
    """创建想法"""
    data = request.get_json() or {}

    # 验证必填字段
    if not data.get('content'):
        return jsonify({'success': False, 'message': '内容不能为空'}), 400

    thought = Thought(
        content=data['content'],
        author=current_user._get_current_object(),
        thought_type=data.get('thought_type', 'note'),
        tags=data.get('tags', ''),
        source_url=data.get('source_url', ''),
        is_public=data.get('is_public', True)
    )

    db.session.add(thought)
    db.session.commit()

    return jsonify({
        'success': True,
        'message': '想法创建成功',
        'thought': {
            'id': thought.id,
            'content': thought.content,
            'thought_type': thought.thought_type,
            'tags': thought.tags,
            'timestamp': thought.timestamp.isoformat(),
            'is_public': thought.is_public,
            'author': thought.author.username,
            'content_html': thought.content_html
        }
    }), 201


@api.route('/thoughts/<int:id>', methods=['PUT'])
@login_required
def update_thought(id):
    """更新想法"""
    thought = Thought.query.get_or_404(id)

    # 检查权限
    if thought.author_id != current_user.id:
        return jsonify({'success': False, 'message': '无权编辑此想法'}), 403

    data = request.get_json() or {}

    if 'content' in data:
        thought.content = data['content']
    if 'thought_type' in data:
        thought.thought_type = data['thought_type']
    if 'tags' in data:
        thought.tags = data['tags']
    if 'source_url' in data:
        thought.source_url = data['source_url']
    if 'is_public' in data:
        thought.is_public = data['is_public']

    db.session.commit()

    return jsonify({
        'success': True,
        'message': '想法更新成功',
        'thought': {
            'id': thought.id,
            'content': thought.content,
            'thought_type': thought.thought_type,
            'tags': thought.tags,
            'timestamp': thought.timestamp.isoformat(),
            'is_public': thought.is_public,
            'author': thought.author.username,
            'content_html': thought.content_html
        }
    })


@api.route('/thoughts/<int:id>', methods=['DELETE'])
@login_required
def delete_thought(id):
    """删除想法（软删除）"""
    thought = Thought.query.get_or_404(id)

    # 检查权限
    if thought.author_id != current_user.id:
        return jsonify({'success': False, 'message': '无权删除此想法'}), 403

    thought.is_deleted = True
    db.session.commit()

    return jsonify({
        'success': True,
        'message': '想法删除成功'
    })


@api.route('/thoughts/search', methods=['GET'])
def search_thoughts():
    """搜索想法"""
    query = request.args.get('q', '').strip()
    if not query:
        return jsonify({'success': False, 'message': '搜索关键词不能为空'}), 400

    page = request.args.get('page', 1, type=int)
    per_page = min(request.args.get('per_page', 20, type=int), 100)

    # 构建搜索查询
    search_query = Thought.query.filter(
        Thought.content.contains(query),
        Thought.is_deleted == False
    )

    # 筛选公开状态
    if not current_user.is_authenticated:
        search_query = search_query.filter_by(is_public=True)
    else:
        # 显示公开想法和用户自己的私有想法
        search_query = search_query.filter(
            (Thought.is_public == True) | (Thought.author_id == current_user.id)
        )

    # 根据类型筛选
    thought_type = request.args.get('type')
    if thought_type:
        search_query = search_query.filter_by(thought_type=thought_type)

    # 排序
    search_query = search_query.order_by(Thought.timestamp.desc())

    pagination = search_query.paginate(page=page, per_page=per_page, error_out=False)

    thoughts = []
    for thought in pagination.items:
        thoughts.append({
            'id': thought.id,
            'content': thought.content,
            'thought_type': thought.thought_type,
            'tags': thought.tags,
            'timestamp': thought.timestamp.isoformat(),
            'is_public': thought.is_public,
            'author': thought.author.username if thought.author else 'Anonymous',
            'content_html': thought.content_html
        })

    return jsonify({
        'success': True,
        'thoughts': thoughts,
        'query': query,
        'pagination': {
            'page': pagination.page,
            'pages': pagination.pages,
            'per_page': pagination.per_page,
            'total': pagination.total,
            'has_prev': pagination.has_prev,
            'has_next': pagination.has_next,
            'prev_num': pagination.prev_num,
            'next_num': pagination.next_num
        }
    })


@api.route('/thoughts/tag/<tag>', methods=['GET'])
def get_thoughts_by_tag(tag):
    """根据标签获取想法"""
    page = request.args.get('page', 1, type=int)
    per_page = min(request.args.get('per_page', 20, type=int), 100)

    query = Thought.query.filter(
        Thought.tags.contains(tag),
        Thought.is_deleted == False
    )

    # 筛选公开状态
    if not current_user.is_authenticated:
        query = query.filter_by(is_public=True)
    else:
        query = query.filter(
            (Thought.is_public == True) | (Thought.author_id == current_user.id)
        )

    pagination = query.order_by(Thought.timestamp.desc())\
                    .paginate(page=page, per_page=per_page, error_out=False)

    thoughts = []
    for thought in pagination.items:
        thoughts.append({
            'id': thought.id,
            'content': thought.content,
            'thought_type': thought.thought_type,
            'tags': thought.tags,
            'timestamp': thought.timestamp.isoformat(),
            'is_public': thought.is_public,
            'author': thought.author.username if thought.author else 'Anonymous',
            'content_html': thought.content_html
        })

    return jsonify({
        'success': True,
        'thoughts': thoughts,
        'tag': tag,
        'pagination': {
            'page': pagination.page,
            'pages': pagination.pages,
            'per_page': pagination.per_page,
            'total': pagination.total,
            'has_prev': pagination.has_prev,
            'has_next': pagination.has_next,
            'prev_num': pagination.prev_num,
            'next_num': pagination.next_num
        }
    })


@api.route('/thoughts/user/<username>', methods=['GET'])
def get_user_thoughts(username):
    """获取指定用户的想法"""
    from app.models import User

    user = User.query.filter_by(username=username).first_or_404()
    page = request.args.get('page', 1, type=int)
    per_page = min(request.args.get('per_page', 20, type=int), 100)

    query = Thought.query.filter_by(author=user, is_deleted=False)

    # 只有用户本人或管理员可以查看私有想法
    if not current_user.is_authenticated or (current_user.username != username and not current_user.can(Permission.ADMIN)):
        query = query.filter_by(is_public=True)

    pagination = query.order_by(Thought.timestamp.desc())\
                    .paginate(page=page, per_page=per_page, error_out=False)

    thoughts = []
    for thought in pagination.items:
        thoughts.append({
            'id': thought.id,
            'content': thought.content,
            'thought_type': thought.thought_type,
            'tags': thought.tags,
            'timestamp': thought.timestamp.isoformat(),
            'is_public': thought.is_public,
            'author': thought.author.username,
            'content_html': thought.content_html
        })

    return jsonify({
        'success': True,
        'thoughts': thoughts,
        'username': username,
        'pagination': {
            'page': pagination.page,
            'pages': pagination.pages,
            'per_page': pagination.per_page,
            'total': pagination.total,
            'has_prev': pagination.has_prev,
            'has_next': pagination.has_next,
            'prev_num': pagination.prev_num,
            'next_num': pagination.next_num
        }
    })


@api.route('/thoughts/<int:id>/toggle-public', methods=['POST'])
@login_required
def toggle_thought_public(id):
    """切换想法的公开状态"""
    thought = Thought.query.get_or_404(id)

    # 检查权限
    if thought.author_id != current_user.id:
        return jsonify({'success': False, 'message': '无权修改此想法'}), 403

    thought.is_public = not thought.is_public
    db.session.commit()

    return jsonify({
        'success': True,
        'message': f'想法已设为{"公开" if thought.is_public else "私有"}',
        'is_public': thought.is_public
    })


@api.route('/thoughts/types', methods=['GET'])
def get_thought_types():
    """获取想法类型列表"""
    types = [
        {'value': 'note', 'label': '笔记', 'color': 'blue'},
        {'value': 'quote', 'label': '引用', 'color': 'green'},
        {'value': 'idea', 'label': '想法', 'color': 'purple'},
        {'value': 'task', 'label': '任务', 'color': 'orange'}
    ]

    return jsonify({
        'success': True,
        'types': types
    })


@api.route('/thoughts/tags/popular', methods=['GET'])
def get_popular_tags():
    """获取热门标签"""
    # 这里简化处理，实际项目中可能需要更复杂的统计逻辑
    thoughts = Thought.query.filter_by(is_deleted=False, is_public=True).all()

    tag_count = {}
    for thought in thoughts:
        if thought.tags:
            tags = [tag.strip() for tag in thought.tags.split(',')]
            for tag in tags:
                if tag:
                    tag_count[tag] = tag_count.get(tag, 0) + 1

    # 按使用次数排序并返回前20个
    popular_tags = sorted(tag_count.items(), key=lambda x: x[1], reverse=True)[:20]

    return jsonify({
        'success': True,
        'tags': [{'tag': tag, 'count': count} for tag, count in popular_tags]
    })